compiler/noirc_evaluator/src/ssa/opt/array_get.rs

```diff
@@ -0,0 +1,156 @@
+use crate::ssa::{
+    ir::{
+        dfg::CallStack, function::Function, instruction::Instruction, map::Id, types::Type,
+        value::ValueId,
+    },
+    ssa_gen::Ssa,
+};
+use fxhash::FxHashMap as HashMap;
+
+impl Ssa {
+    // Given an original IfElse instruction is this:
+    //
+    //     v10 = if v0 then v2 else if v1 then v3
+    //
+    // and a later ArrayGet instruction is this:
+    //
+    //     v11 = array_get v4, index v4
+    //
+    // we optimize it to this:
+    //
+    //     v12 = array_get v2, index v4
+    //     v13 = array_get v3, index v4
+    //     v14 = if v0 then v12 else if v1 then v13
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(crate) fn array_get_optimization(mut self) -> Self {
+        for function in self.functions.values_mut() {
+            // This should match the check in flatten_cfg
+            if let crate::ssa::ir::function::RuntimeType::Brillig = function.runtime() {
+                continue;
+            }
+
+            Context::default().optimize_array_get(function);
+        }
+
+        self
+    }
+}
+
+#[derive(Default)]
+struct Context {
+    // Given an IfElse instruction, here we map its result to that instruction.
+    // We only capture such values if the IfElse values are arrays.
+    result_if_else: HashMap<ValueId, Id<Instruction>>,
+}
+
+impl Context {
+    fn optimize_array_get(&mut self, function: &mut Function) {
+        let block = function.entry_block();
+        let dfg = &mut function.dfg;
+        let instructions = dfg[block].take_instructions();
+
+        for instruction_id in instructions {
+            let instruction = dfg[instruction_id].clone();
+
+            match &instruction {
+                Instruction::IfElse { then_value, .. } => {
+                    let then_value = *then_value;
+
+                    // Only apply this optimization to IfElse where values are arrays
+                    let Type::Array(..) = dfg.type_of_value(then_value) else {
+                        continue;
+                    };
+
+                    let results = dfg.instruction_results(instruction_id);
+                    let result = results[0];
+                    self.result_if_else.insert(result, instruction_id);
+
+                    dfg[block].instructions_mut().push(instruction_id);
+                }
+                Instruction::ArrayGet { array, index } => {
```

Comment:
Note that the optimization is done even if the index is known at compile-time. Maybe that's fine? I think right now if that's the case another optimization takes care of that...

Another thought: should a similar optimization be done with `ArraySet`?

Reply:
> Another thought: should a similar optimization be done with ArraySet?

Potentially but let's restrict this PR to reads to make it easier to think about. We can follow up later if necessary.